// get-image-colors/index.js
const getPixels = require('get-pixels')
const getRgbaPalette = require('get-rgba-palette')
const chroma = require('chroma-js')
const getSvgColors = require('get-svg-colors')
const pify = require('pify')

const patterns = {
  image: /\.(gif|jpg|png|svg)$/i,
  raster: /\.(gif|jpg|png)$/i,
  svg: /svg$/i
}

//added a count to the options part of teh 
function colorPalette (input, count, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {
      type: undefined,
      count: count
    }
  } else if (typeof options === 'string') {
    options = {
      type: options,
      count: count
    }
  }

  // SVG
  if (!Buffer.isBuffer(input)) {
    if (input.match(patterns.svg)) {
      return callback(null, getSvgColors(input, { flat: true }))
    }
  } else if (options.type === 'image/svg+xml') {
    return callback(null, getSvgColors(input, { flat: true }))
  }

  // PNG, GIF, JPG
  return paletteFromBitmap(input, count, options, callback)
}

function paletteFromBitmap (filename, count, options, callback) {
  if (!callback) {
    callback = options
    options = {
      type: undefined,
      count: 5 // not sure if I should change that? I won't be using bitmap anytime soon
    }
  }

  getPixels(filename, options.type, function (err, pixels) {
    if (err) return callback(err)
    const palette = getRgbaPalette(pixels.data, count).map(function (rgba) {
      return chroma(rgba)
    })

    return callback(null, palette)
  })
}

module.exports = pify(colorPalette)
